File: hw1/analytical.py
import pandas as pd

accounts=pd.read_csv('data/accounts.csv')
cards=pd.read_csv('data/cards.csv')
clients=pd.read_csv('data/clients.csv')
districts=pd.read_csv('data/districts.csv')
links=pd.read_csv('data/links.csv')
loans=pd.read_csv('loans_py.csv')
payment_orders=pd.read_csv('data/payment_orders.csv')
transactions=pd.read_csv('data/transactions.csv')

tidy_acc=accounts
tidy_acc.columns=['account_id','district_id','open_date','statement_frequency']

tidy_dis=districts.loc[:,['id','name']]
tidy_dis.columns=['district_id','district_name']

temp=links.groupby('account_id').size()
tidy_links=pd.DataFrame(temp)
tidy_links.columns=['num_customers']
tidy_links['account_id']=temp.index

temp_links=links.loc[:,['client_id','account_id']]
temp_cards=cards.loc[:,['id','link_id']]
temp_cards.columns=['id','client_id']
temp=temp_links.join(temp_cards.set_index('client_id'), on='client_id')

temp=temp.loc[~temp.id.isna(),:]
temp_tab=temp.groupby('account_id').size()
temp_credit=pd.DataFrame(temp_tab)
temp_credit.columns=['credit_cards']
temp_credit['account_id']=temp_credit.index


tidy_loans=loans.drop(columns=['id','date'])
tidy_loans.columns=['account_id','loan_amount','loan_payments','loan_term','loan_status']
tidy_loans['loan_default']=tidy_loans['loan_status']=='B'
has_loan=tidy_acc['account_id'].isin(tidy_loans['account_id'])

tidy_wit=transactions.loc[transactions['type']=='credit',:]
tidy_wit=tidy_wit.loc[:,['account_id','amount']]
wit_max=pd.DataFrame(tidy_wit.groupby('account_id').max())
wit_min=pd.DataFrame(tidy_wit.groupby('account_id').min())
wit_cc=pd.DataFrame(tidy_wit.groupby('account_id').size())
wit_max.columns=['max_withdrawal']
wit_min.columns=['min_withdrawal']
wit_cc.columns=['cc_payments']
wit_max['account_id']=wit_max.index
wit_min['account_id']=wit_min.index
wit_cc['account_id']=wit_cc.index

tidy_bal=transactions.loc[:,['account_id','balance']]
bal_max=pd.DataFrame(tidy_bal.groupby('account_id').max())
bal_min=pd.DataFrame(tidy_bal.groupby('account_id').min())
bal_max.columns=['max_balance']
bal_min.columns=['min_balance']
bal_max['account_id']=bal_max.index
bal_min['account_id']=bal_min.index

final_df=tidy_acc.join(tidy_dis.set_index('district_id'), on='district_id')
final_df=final_df.drop(columns=['district_id'])
final_df=final_df.join(tidy_links.set_index('account_id'), on='account_id')
final_df.loc[final_df['num_customers'].isna(),'num_customers']=0
final_df=final_df.join(temp_credit.set_index('account_id'), on='account_id')
final_df.loc[final_df['credit_cards'].isna(),'credit_cards']=0
final_df['loan']=has_loan
final_df=final_df.join(tidy_loans.set_index('account_id'), on='account_id')
final_df=final_df.join(wit_max.set_index('account_id'), on='account_id')
final_df=final_df.join(wit_min.set_index('account_id'), on='account_id')
final_df=final_df.join(wit_cc.set_index('account_id'), on='account_id')
final_df=final_df.join(bal_max.set_index('account_id'), on='account_id')
final_df=final_df.join(bal_min.set_index('account_id'), on='account_id')

final_df.to_csv('analytical_py.csv',index=False)

File: hw1/analytical.R
library(rstudioapi)
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

library(reshape)
library(dplyr)
library(tidyr)
library(plyr)
library(stringr)

##load data
accounts<-read.csv('data/accounts.csv')
cards<-read.csv('data/cards.csv')
clients<-read.csv('data/clients.csv')
districts<-read.csv('data/districts.csv')
links<-read.csv('data/links.csv')
loans<-read.csv('loans_r.csv')
payment_orders<-read.csv('data/payment_orders.csv')
transactions<-read.csv('data/transactions.csv')

##brief view
head(accounts)
head(districts)
head(cards)
head(clients)
head(links)
head(loans)
head(payment_orders)
head(transactions)


tidy_acc=accounts
colnames(tidy_acc)=c('account_id','district_id','open_date','statement_frequency')

tidy_dis=select(districts,c(id,name))
colnames(tidy_dis)=c('district_id','district_name')

temp=table(links$account_id)

tidy_lin=data.frame(temp)
colnames(tidy_lin)=c('account_id','num_customers')

temp_links=links[,c('client_id','account_id')]
temp_cards=cards[,c('id','link_id')]
colnames(temp_cards)=c('id','client_id')
temp=join(temp_links,temp_cards, by='client_id')


temp=filter(temp,id!='NA')
temp_tab=table(temp$account_id)
temp_cards=as.data.frame(temp_tab)
colnames(temp_cards)=c('account_id','credit_cards')
temp_cards$account_id=as.integer(as.character(temp_cards$account_id))

tidy_loans=select(loans,-c(id,date))
colnames(tidy_loans)=c('account_id','loan_amount','loan_payments',
                       'loan_term','loan_status')
tidy_loans['loan_default']=tidy_loans$loan_status=='B'
has_loan=final_df$account_id %in% tidy_loans$account_id

tidy_wit=select(filter(transactions,type=='credit'),c(account_id,amount))
wit_max=aggregate(amount~account_id,data=tidy_wit,max)
wit_min=aggregate(amount~account_id,data=tidy_wit,min)
wit_cc=table(tidy_wit$account_id)
wit_cc=as.data.frame(wit_cc)
colnames(wit_max)=c('account_id','max_withdrawal')
colnames(wit_min)=c('account_id','min_withdrawal')
colnames(wit_cc)=c('account_id','cc_payments')

tidy_bal=select(transactions,c(account_id,balance))
bal_max=aggregate(balance~account_id,data=tidy_bal,max)
bal_min=aggregate(balance~account_id,data=tidy_bal,min)
colnames(bal_max)=c('account_id','max_balance')
colnames(bal_min)=c('account_id','min_balance')

final_df=join(tidy_acc,tidy_dis,by='district_id')
final_df=select(final_df,-c(district_id))
final_df=join(final_df,tidy_lin,by='account_id')
final_df$num_customers[is.na(final_df$num_customers)]=0
final_df=join(final_df,temp_cards,by='account_id')
final_df$credit_cards[is.na(final_df$credit_cards)]=0
final_df['loan']=has_loan
final_df=join(final_df,tidy_loans,by='account_id')
final_df=join(final_df,wit_max,by='account_id')
final_df=join(final_df,wit_min,by='account_id')
final_df=join(final_df,wit_cc,by='account_id')
final_df=join(final_df,bal_max,by='account_id')
final_df=join(final_df,bal_min,by='account_id')

colnames(final_df)

write.csv(final_df,'analytical_r.csv',row.names=FALSE)
